test/test_elastic_driver.py

```diff
@@ -0,0 +1,351 @@
+            assert updated_slot_info.rank == slot_info.rank % 2, rank
+            assert updated_slot_info.local_size == slot_info.local_size, rank
+            assert updated_slot_info.local_rank == slot_info.local_rank, rank
+            assert updated_slot_info.cross_size == 1, rank
+            assert updated_slot_info.cross_rank == 0, rank
+
+    @mock.patch('horovod.run.elastic.driver.DISCOVER_HOSTS_FREQUENCY_SECS', 0.01)
+    def test_rank_and_size_with_host_added(self):
+        """Tests training starts with one host two slots, then a second host is added."""
+        slots = {'host-1': 2}
+        discovery = FixedHosts(slots)
+
+        def add_host():
+            slots = {'host-1': 2, 'host-2': 2}
+            discovery.set(slots)
+
+        driver = ElasticDriver(mock.Mock(), discovery, min_np=2, max_np=4)
+        driver.wait_for_available_hosts(min_np=2)
+
+        rank_results = {}
+
+        def exec_command(slot_info, events):
+            driver.record_ready(slot_info.hostname, slot_info.local_rank)
+
+            if slot_info.hostname == 'host-1':
+                if slot_info.rank == 0:
+                    add_host()
+                driver.wait_for_available_hosts(4)
+                driver.record_ready(slot_info.hostname, slot_info.local_rank)
+
+            driver.record_ready(slot_info.hostname, slot_info.local_rank)
+            updated_slot_info = driver.get_slot_info(slot_info.hostname, slot_info.local_rank)
+            rank_results[slot_info.rank] = (slot_info, updated_slot_info)
+            return 0, time.time()
+
+        driver.start(np=2, create_worker_fn=exec_command)
+        res = driver.get_results()
+        driver.stop()
+
+        assert len(res) == 4
+        for name, (exit_code, timestamp) in res.items():
+            assert exit_code == 0, name
+
+        assert len(rank_results) == 4
+        for rank, (slot_info, updated_slot_info) in rank_results.items():
+            assert updated_slot_info.size == 4, rank
+            assert updated_slot_info.rank == slot_info.rank, rank
+            assert updated_slot_info.local_size == slot_info.local_size, rank
+            assert updated_slot_info.local_rank == slot_info.local_rank, rank
+            assert updated_slot_info.cross_size == 2, rank
+            assert updated_slot_info.cross_rank == slot_info.cross_rank, rank
+
+    @mock.patch('horovod.run.elastic.driver.DISCOVER_HOSTS_FREQUENCY_SECS', 0.01)
+    def test_wait_for_available_hosts(self):
+        """Tests that driver blocks until the min number of slots are available."""
+        slots = [{'host-1': 4},
+                 {'host-1': 4, 'host-2': 8},
+                 {'host-1': 4, 'host-2': 8, 'host-3': 4}]
+        discovery = HostDiscoverySequence(slots)
+
+        driver = ElasticDriver(mock.Mock(), discovery, min_np=2, max_np=12)
+        driver.wait_for_available_hosts(min_np=10)
+        assert driver._host_manager.current_hosts.count_available_slots() >= 10
```

Comment:
Can we assert here that `wait_for_available_hosts` returned on the second slots sequence, not the third one? I guess this is hard because discovery polls the sequences in a separate thread at high frequency. Even the returned `current_hosts` might come from the third slots sequence if thread is too quick.

Reply:
Usually it returns on the third sequence just because the threads progress at different rates.  We could test this behavior by removing the third sequence, but the behavior is the same.